### PR TITLE
monitoring: move to OTEL sync gauge

### PR DESCRIPTION
**Pull request by howardjohn**
Added in https://github.com/open-telemetry/opentelemetry-go/pull/5304.

Prior to this, we hack around the lack of a sync gauge by approximating one with an async gauge. This is complex and has a race condition, captured by the test I added.

If we compare the performance:

```
name                     old time/op    new time/op    delta
Gauge/no_labels-16         13.0ns ± 1%    50.7ns ± 2%  +290.07%  (p=0.004 n=8+4)
Gauge/dynamic_labels-16     322ns ± 7%     385ns ± 7%   +19.54%  (p=0.002 n=10+4)
Gauge/static_labels-16     13.2ns ± 9%   117.2ns ± 3%  +788.01%  (p=0.004 n=8+4)

name                     old alloc/op   new alloc/op   delta
Gauge/no_labels-16          0.00B          0.00B           ~     (all equal)
Gauge/dynamic_labels-16      288B ± 0%      320B ± 0%   +11.11%  (p=0.002 n=10+4)
Gauge/static_labels-16      0.00B          0.00B           ~     (all equal)

name                     old allocs/op  new allocs/op  delta
Gauge/no_labels-16           0.00           0.00           ~     (all equal)
Gauge/dynamic_labels-16      4.00 ± 0%      6.00 ± 0%   +50.00%  (p=0.002 n=10+4)
Gauge/static_labels-16       0.00           0.00           ~     (all equal)
```

We see a regression. However, this brings it in line with Sum:

```
BenchmarkCounter
BenchmarkCounter/no_labels
BenchmarkCounter/no_labels-16           22860031                49.13 ns/op            0 B/op          0 allocs/op
BenchmarkCounter/dynamic_labels
BenchmarkCounter/dynamic_labels-16       3263858               421.3 ns/op           320 B/op          6 allocs/op
BenchmarkCounter/static_labels
BenchmarkCounter/static_labels-16        8653600               128.8 ns/op             0 B/op          0 allocs/op
BenchmarkGauge
BenchmarkGauge/no_labels
BenchmarkGauge/no_labels-16             23151216                47.95 ns/op            0 B/op          0 allocs/op
BenchmarkGauge/dynamic_labels
BenchmarkGauge/dynamic_labels-16         3267243               370.7 ns/op           320 B/op          6 allocs/op
BenchmarkGauge/static_labels
BenchmarkGauge/static_labels-16         10233974               112.9 ns/op             0 B/op          0 allocs/op
```

Which I think is acceptable. In fact, the only reason we had such strong performance was by bypassing safety-critical locks

### CODE REVIEW

File: pkg/monitoring/gauge.go [hunk 1]
```diff
@@ -65,40 +48,23 @@ func newGauge(o options) *gauge {
...
 func (f *gauge) Record(value float64) {
 	f.runRecordHook(value)
-	// TODO: https://github.com/open-telemetry/opentelemetry-specification/issues/2318 use synchronous gauge so we don't need to deal with this
-	f.attributeSetsMutex.Lock()
-	// Special case: we lazy-load the non-labeled value. This ensures that metrics which should always have labels do not end up with a un-labeled zero-value
-	// If a metric really requires `metric{} 0`, they can explicitly call .Record(0).
-	if f.currentGaugeSet == nil {
-		f.currentGaugeSet = &gaugeValues{}
-		f.attributeSets[attribute.NewSet()] = f.currentGaugeSet
+	if f.precomputedRecordOption != nil {
```
Comment (wulianglongrd) [inline pkg/monitoring/gauge.go:51]: Can this check be deleted?

Comment (howardjohn) [reply pkg/monitoring/gauge.go:51]: No, this is only used after With() is called. You can have no options

Comment (wulianglongrd) [reply pkg/monitoring/gauge.go:51]: I mean when `f.precomputedRecordOption == nil`, `f.precomputedRecordOption...` is also ok.

Comment (howardjohn) [reply pkg/monitoring/gauge.go:51]: I see, I might haev done it to reduce allocations. let me check

Comment (howardjohn) [reply pkg/monitoring/gauge.go:51]: Nope... good call

**Review by lei-tang — APPROVED** [review]

**Pull request merged**